### PR TITLE
[docs] Add new cursor-based pagination paragraph

### LINKED ISSUE #1906: [DataGrid] Improve support of cursor-based pagination
- [x] I have searched the [issues](https://github.com/mui-org/material-ui-x/issues) of this repository and believe that this is not a duplicate.

## Summary 💡

Hi all,

Thanks for your work with this great library - the `DataGrid` component is a bliss to work with. On the same note, I stumbled upon an improvement idea in regards to cursor-based pagination.

At the moment, the `DataGrid` API mainly caters to page-based pagination using page numbers. For this a consumer of the component would use the `onPageChange` handler to re-fetch data from the API using the new page number the user navigated to. While this works for page-based pagination, it's less ideal for cursor-based pagination as we don't know (without additional checks) whether the user has navigate forward or backward in the `DataGrid` `onPageChange` handler.

**Hence, I propose that we create two additional event handlers:**
- `onNextPageChange` - called when a user navigates to the next page
- `onPreviousPageChange` - called when a user navigates to the previous page

#### Before

```tsx
  const [currentPageIndex, setCurrentPageIndex] = useState(0)
  const { data, fetchNextPage, fetchPreviousPage } = useSomeData(pageSize) // fetching data using react-query

  const handlePageChange = (params: GridPageChangeParams) => {
    // If we navigate FORWARD in our pages, i.e. the new page number is higher than current page
    if (params.page > currentPageIndex) {
      fetchNextPage()
    }
    // If we navigate BACKWARD in our pages, i.e. the new page number is lower than current page
    else {
      fetchPreviousPage()
    }
    setCurrentPageIndex(params.page)
  }

  const rows = data?.pages[currentPageIndex]?.edges?.map(({ node }) => node)

  return (
    <Page>
      {rows && rows.length > 0 ? (
        <Card>
          <DataGrid
            autoHeight
            pagination
            pageSize={10}
            page={currentPageIndex}
            rowCount={data?.pages[currentPageIndex]?.totalCount}
            paginationMode="server"
            rowsPerPageOptions={[10, 20, 50]}
            onPageChange={handlePageChange}
            rows={rows}
            columns={columns}
          />
        </Card>
      ) : null}
    </Page>
  )
```

#### After

```tsx
  const [currentPageIndex, setCurrentPageIndex] = useState(0)
  const { data, fetchNextPage, fetchPreviousPage } = useSomeData(pageSize) // fetching data using react-query

  const rows = data?.pages[currentPageIndex]?.edges?.map(({ node }) => node)

  return (
    <Page>
      {rows && rows.length > 0 ? (
        <Card>
          <DataGrid
            autoHeight
            pagination
            pageSize={10}
            page={currentPageIndex}
            rowCount={data?.pages[currentPageIndex]?.totalCount}
            paginationMode="server"
            rowsPerPageOptions={[10, 20, 50]}
            onPageChange={(param: GridPageChangeParams) => setCurrentPageIndex(params.page)}
            onNextPageChange={fetchNextPage}
            onPreviousPageChange={fetchPreviousPage}
            rows={rows}
            columns={columns}
          />
        </Card>
      ) : null}
    </Page>
  )
```

## Motivation 🔦

This addition would make it a bit easier to use the `DataGrid` component for cursor-based pagination.

### PR DESCRIPTION
**Pull request by flaviendelangle**
Closes #1906 

Preview: https://deploy-preview-1991--material-ui-x.netlify.app/components/data-grid/pagination/#cursor-based-pagination

### CODE REVIEW

File: docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    [response.data],
+  );
+
+  return {
+    ...response,
+    data,
+  };
+};
+
+export default function ServerPaginationGrid() {
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx:55]:
```suggestion
export default function CursorPaginationGrid() {
```

File: docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  const pagesNextCursor = React.useRef<{ [page: number]: string }>({});
+
+  const [rows, setRows] = React.useState<GridRowsProp>([]);
+  const [page, setPage] = React.useState(0);
+  const [loading, setLoading] = React.useState<boolean>(false);
+
+  const handlePageChange = (params: GridPageChangeParams) => {
+    setPage(params.page);
+  };
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx:70]: We might not want to allow going to a page where we don't have loaded yet the next page cursor. Meaning, if you go quickly from page 0 to page 2. We have no ways to load the data, as we need page 1 to now the next cursor.
```suggestion
  const handlePageChange = (params: GridPageChangeParams) => {
    // We have the cursor, we can allow the page transition.
    if (pagesNextCursor.current[params.page - 1]) {
      setPage(params.page);
    }
  };
```

Comment (oliviertassinari) [reply docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx:70]: @tobiasbueschel Would this new demo be enough for your use case? (we won't change the API just yet). How do you solve this problem with react-query?

File: docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.js [hunk 3]
```diff
@@ -59,8 +59,6 @@ export default function CursorPaginationGrid() {
     (async () => {
       const nextCursor = pagesNextCursor.current[page - 1];
 
-      console.log(page);
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.js]: Use the toggle to work with the TS demo directly:

Capture d’écran 2021-07-01 à 02 15 28

File: docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  return (
+    <div style={{ height: 400, width: '100%' }}>
+      <DataGrid
+        rows={rows}
+        columns={data.columns}
+        pagination
+        pageSize={5}
+        rowCount={100}
+        paginationMode="server"
+        onPageChange={handlePageChange}
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx:113]: We need to force the page to sync the UI with the state.

```suggestion
        page={page}
        onPageChange={handlePageChange}
```

**Review by oliviertassinari — COMMENTED** [review]
Looks almost good to go

File: docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+const PAGE_SIZE = 5;
+
+function loadServerRows(
+  cursor: string | null | undefined,
+  data: CursorBasedGridData,
+): Promise<ServerBasedGridResponse> {
+  return new Promise<ServerBasedGridResponse>((resolve) => {
+    setTimeout(() => {
+      const start = cursor ? data.cursors.indexOf(cursor) + 1 : 0;
```
Comment (oliviertassinari) [inline docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.tsx:31]: Off by 1 computation error

```suggestion
      const start = cursor ? data.cursors.indexOf(cursor) : 0;
```

**Review by oliviertassinari — COMMENTED** [review]
I can't notice anything else to improve

File: docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.js [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+  });
+}
+
+const useCursorBasedDemoData = (options) => {
+  const response = useDemoData(options);
+
+  const data = React.useMemo(
+    () => ({
+      ...response.data,
+      cursors: response.data.rows.map(() => Math.random().toString()),
```
Comment (dtassone) [inline docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.js:25]: We could `row.Id` as cursor

Comment (dtassone) [reply docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.js:25]: https://slack.engineering/evolving-api-pagination-at-slack/

Comment (flaviendelangle) [reply docs/src/pages/components/data-grid/pagination/CursorPaginationGrid.js]: Done :+1:

**Review by dtassone — APPROVED** [review]

**Comment by oliviertassinari** [comment]
Great 👍

**Pull request merged**